### PR TITLE
lib: net_buf: move the network buffer implementation to lib

### LINKED ISSUE #36374: Rename net_buf to something more generic and move it outside of net/
Networking buffers is a separate subsystem with no dependency on networking itself. This implementation was originally created for Bluetooth and called bt_buf back then. Once work started to create a networking stack for Zephyr it made sense to reuse the buffer implementation, but Networking depending on something called bt_buf didn't feel right. Consequently, the implementation was renamed to net_buf and moved under net/ due to its namespace.

There are a couple of issues with the current situation:

 - There is no direct dependency on the networking stack, but the naming gives an impression that there might be
 - The API is used by other subsystems (a major one being Bluetooth) even though these also don't have any dependency on the networking stack

It would therefore make sense to consider renaming the subsystem and API to something more generic, and the natural choice for that would be sys_buf (based on current Zephyr naming guidelines).

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Move the net_buf implementation from the networking subsystem to a library as they have no dependency on the networking subsystem.

Network buffers are used in subsystems outside of networking (e.g. Bluetooth, USB).

Fixes: #36374

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

### CODE REVIEW

File: doc/connectivity/networking/api/buf_mgmt.rst [hunk 1]
```diff
@@ -1,10 +1,9 @@
 .. _net_buf_mgmt:
 
-Network Buffer Management
+Network Packet Management
```
Comment (Thalley) [inline doc/connectivity/networking/api/buf_mgmt.rst:3]: With this rename, the file name seems a bit odd. Should it be renamed to `buf_pkt.rst`? Ditto with the label

Comment (jukkar) [reply doc/connectivity/networking/api/buf_mgmt.rst:3]: The label could be changed to `_net_pkt_mgmt`. I am ok with current filename, but if we want to change it, the proper name could perhaps be `pkt_mgmt.rst` or `net_pkt_mgmt.rst`

Comment (henrikbrixandersen) [reply doc/connectivity/networking/api/buf_mgmt.rst:3]: Thank you for reviewing. I have renamed the file and changed the label as suggested above.

**Comment by henrikbrixandersen** [comment]
I've split the addition of the `include/zephyr/net/buf.h` compatibility header to its own commit to preserve `git blame` on `include/zephyr/net_buf.h`.

This shouldn't affect Git bisect, as all header usage was updated in the commit renaming the header.

**Review by rlubos — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]
Nice work!

Did you have some thoughts/plan for renaming the API as the next step (regarding the new name as well as some graceful migration solution)?

**Comment by jhedberg** [comment]
> Did you have some thoughts/plan for renaming the API as the next step (regarding the new name as well as some graceful migration solution)?

Regarding this next step, I think ideally we should do it all for Zephyr 4.0, so that we don't end up causing trouble for out-of-tree users for two releases in a row.

**Comment by henrikbrixandersen** [comment]
> Nice work!

Thanks!

> Did you have some thoughts/plan for renaming the API as the next step (regarding the new name as well as some graceful migration solution)?

I kind of like the "Network Buffer" name, but we could rename it to "sys_buf" or similar, creating a new `include/zephyr/sys_buf.h` header file with the renamed API functions, and change `include/zephyr/net_buf.h` to be deprecated, static inline functions calling their `sys_buf_*` counterparts.

I can take of that as a step 2 (separate PR), but we'd need consensus on this first (there was no clear consensus in https://github.com/zephyrproject-rtos/zephyr/issues/36374).

**Comment by henrikbrixandersen** [comment]
> Regarding this next step, I think ideally we should do it all for Zephyr 4.0, so that we don't end up causing trouble for out-of-tree users for two releases in a row.

Agreed. If we plan on this, we should do the rename well before v4.0.0.

**Review by Thalley — DISMISSED** [review]

**Comment by jukkar** [comment]
> I kind of like the "Network Buffer" name, but we could rename it to "sys_buf" or similar, creating a new `include/zephyr/sys_buf.h` header file with the renamed API functions, and change `include/zephyr/net_buf.h` to be deprecated, static inline functions calling their `sys_buf_*` counterparts.
> 
> I can take of that as a step 2 (separate PR), but we'd need consensus on this first (there was no clear consensus in #36374).

Is there really a need to rename to `sys_buf`, I mean the net_buf seems to be always related to network related data. But anyway, if there is a desire to change the name, I would like to avoid deprecating the old name. It is just pain when backporting stuff.

**Review by bjarki-andreasen — DISMISSED** [review]

File: doc/connectivity/networking/api/net_pkt_mgmt.rst [hunk 2]
```diff
@@ -1,10 +1,9 @@
-.. _net_buf_mgmt:
+.. _net_pkt_mgmt:
 
-Network Buffer Management
+Network Packet Management
```
Comment (nashif) [inline doc/connectivity/networking/api/net_pkt_mgmt.rst:3]: didn't we agree to remove 'Network' from the name in documentation?

> Agreement to promote it to a general library but keep the existing name temporarily for compatibility and adapt documentation and references, ie. remove networking from the name.

Comment (henrikbrixandersen) [reply doc/connectivity/networking/api/net_pkt_mgmt.rst:3]: I'd prefer to leaving naming intact for now and have a discussion/consensus on the full naming issue (API and description) before doing a follow up PR.

Comment (nashif) [reply doc/connectivity/networking/api/net_pkt_mgmt.rst:3]: This is something that was discussed and agreed on. Removing the 'Network' from the name will be less confusing and forward looking given that it now will live outside of networking. Do not see why this should be second step that we still have to discuss and agree on, see https://github.com/zephyrproject-rtos/zephyr/issues/36374#issuecomment-2123082283 for the minutes.

I am not talking about renaming the API and signatures which can happen later, this is about just dropping 'Networking' from the title of this API.

Does anyone disagree with that?

Comment (henrikbrixandersen) [reply doc/connectivity/networking/api/net_pkt_mgmt.rst:3]: > I am not talking about renaming the API and signatures which can happen later, this is about just dropping 'Networking' from the title of this API.

So you would just call the API "Buffers"?

Comment (nashif) [reply doc/connectivity/networking/api/net_pkt_mgmt.rst:3]: Buffy....

**Review by nashif — DISMISSED** [review]
approving with assumption that renaming of the API and documentation changes to make it less network/bluetooth agnostic are to follow.

**Comment by henrikbrixandersen** [comment]
Rebased on `main` to resolve merge conflict in `doc/releases/migration-guide-4.0.rst`.

**Review by jhedberg — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

**Pull request merged**